### ISSUE TITLE
Differences with Enter-PSSession

### ISSUE THREAD

**Issue by malthe** [jborean93/pypsrp#137]
When using `pypsrp` to start a session, there are some different to a session started with `Enter-PSSession`.

For example, using the `Set-PSDebug` or `Trace-Command` cmdlets seems to have no effect. And similarly, output from a script in a JEA role capabilities file configured using `ScriptsToProcess` does not come through to any of the session streams – such as the output from `Write-Host "Hello world"`.

Is this a scoping issue or some other session configuration option that I have overlooked?

**Comment by jborean93** [comment]
> For example, using the Set-PSDebug or Trace-Command cmdlets seems to have no effect

This is not something I've looked into too much but I've noticed that `Enter-PSSession` always sends a unique command on starting the session that seems related to `Set-PSDebug`.  Essentially it's running a pipeline for `__Set-PSDebugMode -Mode LocalScript, RemoteScript`

```python
ps = PowerShell(rp)
ps.add_cmdlet('__Set-PSDebugMode').add_paramter('Mode', 6) # LocalScript, RemoteScript
ps.invoke()
```

You can see this first message when monitoring the PSRP traffic

<details>
  <summary>Click to expand!</summary>

```xml
Type      : Data
PSGuid    : 46026018-c3e5-425e-8504-45f378738b2d
Stream    : Default
Fragments : ObjectID   : 3
            FragmentID : 0
            Start      : True
            End        : True
            Length     : 2461
Messages  : ObjectID    : 3
            Destination : Server
            MessageType : CREATE_PIPELINE
            RPID        : fdb26075-37b5-4013-9624-30d1143051b3
            PID         : 46026018-c3e5-425e-8504-45f378738b2d
            Object      : 
<Obj RefId="0">
  <MS>
    <Obj N="PowerShell" RefId="1">
      <MS>
        <Obj N="Cmds" RefId="2">
          <TN RefId="0">
            <T>System.Collections.Generic.List`1[[System.Management.Automation.PSObject, System.Management.Automation, Version=7.2.1.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]</T>
            <T>System.Object</T>
          </TN>
          <LST>
            <Obj RefId="3">
              <MS>
                <S N="Cmd">__Set-PSDebugMode</S>
                <B N="IsScript">false</B>
                <Nil N="UseLocalScope" />
                <Obj N="MergeMyResult" RefId="4">
                  <TN RefId="1">
                    <T>System.Management.Automation.Runspaces.PipelineResultTypes</T>
                    <T>System.Enum</T>
                    <T>System.ValueType</T>
                    <T>System.Object</T>
                  </TN>
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeToResult" RefId="5">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergePreviousResults" RefId="6">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeError" RefId="7">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeWarning" RefId="8">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeVerbose" RefId="9">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeDebug" RefId="10">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeInformation" RefId="11">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="Args" RefId="12">
                  <TNRef RefId="0" />
                  <LST>
                    <Obj RefId="13">
                      <MS>
                        <S N="N">Mode</S>
                        <Obj N="V" RefId="14">
                          <TN RefId="2">
                            <T>System.Management.Automation.DebugModes</T>
                            <T>System.Enum</T>
                            <T>System.ValueType</T>
                            <T>System.Object</T>
                          </TN>
                          <ToString>LocalScript, RemoteScript</ToString>
                          <I32>6</I32>
                        </Obj>
                      </MS>
                    </Obj>
                  </LST>
                </Obj>
              </MS>
            </Obj>
          </LST>
        </Obj>
        <B N="IsNested">false</B>
        <Nil N="History" />
        <B N="RedirectShellErrorOutputPipe">true</B>
      </MS>
    </Obj>
    <B N="NoInput">true</B>
    <Obj N="ApartmentState" RefId="15">
      <TN RefId="3">
        <T>System.Threading.ApartmentState</T>
        <T>System.Enum</T>
        <T>System.ValueType</T>
        <T>System.Object</T>
      </TN>
      <ToString>Unknown</ToString>
      <I32>2</I32>
    </Obj>
    <Obj N="RemoteStreamOptions" RefId="16">
      <TN RefId="4">
        <T>System.Management.Automation.RemoteStreamOptions</T>
        <T>System.Enum</T>
        <T>System.ValueType</T>
        <T>System.Object</T>
      </TN>
      <ToString>AddInvocationInfo</ToString>
      <I32>15</I32>
    </Obj>
    <B N="AddToHistory">false</B>
    <Obj N="HostInfo" RefId="17">
      <MS>
        <B N="_isHostUINull">true</B>
        <B N="_isHostRawUINull">true</B>
        <B N="_isHostNull">true</B>
        <B N="_useRunspaceHost">true</B>
      </MS>
    </Obj>
    <B N="IsNested">false</B>
  </MS>
</Obj>
```

</details>

I believe this all ties into the debug extension to PSRP documented at https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-psrdp/0c6174b8-d4d0-4702-953a-dc1b7e696634. Essentially in this mode any debug events come through under a `USER_EVENT` message stored under `rp.user_event`. Currently there is no real notification that this comes through so there's no real easy way to handle it with `pypsrp` right now.

As for `Trace-Command` I believe that links to the point below.

> And similarly, output from a script in a JEA role capabilities file configured using ScriptsToProcess does not come through to any of the session streams – such as the output from Write-Host "Hello world".

I cannot 100% confirm this is your problem with your JEA role but I assume it's related but for `Write-Host` the problem is this is not part of a PowerShell stream. `Write-Host` will call `$host.UI.WriteLine` which is part of a host call and if there is no host associated with the pipeline or runspace then nothing will happen. For example these are the messages that are sent from the server to the client with both `Trace-Command` and `Write-Host`

<details>
  <summary>Click to expand!</summary>

```xml
Type      : Data
PSGuid    : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
Stream    : Default
Fragments : ObjectID   : 6
            FragmentID : 0
            Start      : True
            End        : True
            Length     : 3199
Messages  : ObjectID    : 6
            Destination : Server
            MessageType : CREATE_PIPELINE
            RPID        : fdb26075-37b5-4013-9624-30d1143051b3
            PID         : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
            Object      : 
<Obj RefId="0">
  <MS>
    <Obj N="PowerShell" RefId="1">
      <MS>
        <Obj N="Cmds" RefId="2">
          <TN RefId="0">
            <T>System.Collections.Generic.List`1[[System.Management.Automation.PSObject, System.Management.Automation, Version=7.2.1.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]</T>
            <T>System.Object</T>
          </TN>
          <LST>
            <Obj RefId="3">
              <MS>
                <S N="Cmd">write-host</S>
                <B N="IsScript">false</B>
                <Nil N="UseLocalScope" />
                <Obj N="MergeMyResult" RefId="4">
                  <TN RefId="1">
                    <T>System.Management.Automation.Runspaces.PipelineResultTypes</T>
                    <T>System.Enum</T>
                    <T>System.ValueType</T>
                    <T>System.Object</T>
                  </TN>
                  <ToString>Error</ToString>
                  <I32>2</I32>
                </Obj>
                <Obj N="MergeToResult" RefId="5">
                  <TNRef RefId="1" />
                  <ToString>Output</ToString>
                  <I32>1</I32>
                </Obj>
                <Obj N="MergePreviousResults" RefId="6">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeError" RefId="7">
                  <TNRef RefId="1" />
                  <ToString>Output</ToString>
                  <I32>1</I32>
                </Obj>
                <Obj N="MergeWarning" RefId="8">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeVerbose" RefId="9">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeDebug" RefId="10">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeInformation" RefId="11">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="Args" RefId="12">
                  <TNRef RefId="0" />
                  <LST>
                    <Obj RefId="13">
                      <MS>
                        <Nil N="N" />
                        <S N="V">test</S>
                      </MS>
                    </Obj>
                  </LST>
                </Obj>
              </MS>
            </Obj>
            <Obj RefId="14">
              <MS>
                <S N="Cmd">Out-Default</S>
                <B N="IsScript">false</B>
                <B N="UseLocalScope">true</B>
                <Obj N="MergeMyResult" RefId="15">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeToResult" RefId="16">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergePreviousResults" RefId="17">
                  <TNRef RefId="1" />
                  <ToString>Warning</ToString>
                  <I32>3</I32>
                </Obj>
                <Obj N="MergeError" RefId="18">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeWarning" RefId="19">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeVerbose" RefId="20">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeDebug" RefId="21">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="MergeInformation" RefId="22">
                  <TNRef RefId="1" />
                  <ToString>None</ToString>
                  <I32>0</I32>
                </Obj>
                <Obj N="Args" RefId="23">
                  <TNRef RefId="0" />
                  <LST />
                </Obj>
              </MS>
            </Obj>
          </LST>
        </Obj>
        <B N="IsNested">false</B>
        <S N="History">write-host test</S>
        <B N="RedirectShellErrorOutputPipe">false</B>
      </MS>
    </Obj>
    <B N="NoInput">true</B>
    <Obj N="ApartmentState" RefId="24">
      <TN RefId="2">
        <T>System.Threading.ApartmentState</T>
        <T>System.Enum</T>
        <T>System.ValueType</T>
        <T>System.Object</T>
      </TN>
      <ToString>Unknown</ToString>
      <I32>2</I32>
    </Obj>
    <Obj N="RemoteStreamOptions" RefId="25">
      <TN RefId="3">
        <T>System.Management.Automation.RemoteStreamOptions</T>
        <T>System.Enum</T>
        <T>System.ValueType</T>
        <T>System.Object</T>
      </TN>
      <ToString>0</ToString>
      <I32>0</I32>
    </Obj>
    <B N="AddToHistory">true</B>
    <Obj N="HostInfo" RefId="26">
      <MS>
        <B N="_isHostUINull">true</B>
        <B N="_isHostRawUINull">true</B>
        <B N="_isHostNull">true</B>
        <B N="_useRunspaceHost">true</B>
      </MS>
    </Obj>
    <B N="IsNested">false</B>
  </MS>
</Obj>

Type      : DataAck
PSGuid    : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
Stream    : 
Fragments : 
Messages  : 

Type      : Data
PSGuid    : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
Stream    : Default
Fragments : ObjectID   : 11
            FragmentID : 0
            Start      : True
            End        : True
            Length     : 839
Messages  : ObjectID    : 11
            Destination : Client
            MessageType : INFORMATION_RECORD
            RPID        : fdb26075-37b5-4013-9624-30d1143051b3
            PID         : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
            Object      : 
<Obj RefId="0">
  <MS>
    <Obj N="MessageData" RefId="1">
      <TN RefId="0">
        <T>System.Management.Automation.HostInformationMessage</T>
        <T>System.Object</T>
      </TN>
      <ToString>test</ToString>
      <Props>
        <S N="Message">test</S>
        <B N="NoNewLine">false</B>
        <S N="ForegroundColor">-1</S>
        <S N="BackgroundColor">-1</S>
      </Props>
    </Obj>
    <S N="Source">Write-Host</S>
    <DT N="TimeGenerated">2022-01-13T20:44:40.1790163+10:00</DT>
    <Obj N="Tags" RefId="2">
      <TN RefId="1">
        <T>System.Collections.Generic.List`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]</T>
        <T>System.Object</T>
      </TN>
      <LST>
        <S>PSHOST</S>
      </LST>
    </Obj>
    <S N="User">jborean</S>
    <S N="Computer">jborean-laptop</S>
    <U32 N="ProcessId">213194</U32>
    <U32 N="NativeThreadId">213206</U32>
    <U32 N="ManagedThreadId">6</U32>
  </MS>
</Obj>

Type      : Data
PSGuid    : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
Stream    : Default
Fragments : ObjectID   : 12
            FragmentID : 0
            Start      : True
            End        : True
            Length     : 460
Messages  : ObjectID    : 12
            Destination : Client
            MessageType : PIPELINE_HOST_CALL
            RPID        : fdb26075-37b5-4013-9624-30d1143051b3
            PID         : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
            Object      : 
<Obj RefId="0">
  <MS>
    <I64 N="ci">-100</I64>
    <Obj N="mi" RefId="1">
      <TN RefId="0">
        <T>System.Management.Automation.Remoting.RemoteHostMethodId</T>
        <T>System.Enum</T>
        <T>System.ValueType</T>
        <T>System.Object</T>
      </TN>
      <ToString>WriteLine3</ToString>
      <I32>17</I32>
    </Obj>
    <Obj N="mp" RefId="2">
      <TN RefId="1">
        <T>System.Collections.ArrayList</T>
        <T>System.Object</T>
      </TN>
      <LST>
        <I32>-1</I32>
        <I32>-1</I32>
        <S>test</S>
      </LST>
    </Obj>
  </MS>
</Obj>

Type      : Data
PSGuid    : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
Stream    : Default
Fragments : ObjectID   : 13
            FragmentID : 0
            Start      : True
            End        : True
            Length     : 103
Messages  : ObjectID    : 13
            Destination : Client
            MessageType : PIPELINE_STATE
            RPID        : fdb26075-37b5-4013-9624-30d1143051b3
            PID         : d6039bfd-29ec-4e29-a366-510dbd9a3dbc
            Object      : 
<Obj RefId="0">
  <MS>
    <I32 N="PipelineState">4</I32>
  </MS>
</Obj>
```

</details>

In this case the host call comes through the information stream (addition in v5) but it also makes a `PIPELINE_HOST_CALL` for one of the `$host.UI.WriteLine()` overloads. If not host was present then the server will not send this host call and thus the data is lost.

Now to solve this problem you effectively need to provide your own `PSHost` that can record this data. There is a pre-canned host you can use or inherit from if you wish to provide your own methods

```python
from pypsrp.host import PSHost, PSHostUserInterface
from pypsrp.powershell import RunspacePool, PowerShell
from pypsrp.wsman import WSMan

host_ui = PSHostUserInterface()
host = PSHost(None, None, False, "PSRP Host", None,
              host_ui, None)

with WSMan('server') as wsman, RunspacePool(wsman, host=host) as rp:
    ps = PowerShell(rp)
    ps.add_script('Write-Host test')
    ps.invoke()

print(host_ui.stdout)
```

Before careful with this as the host implementation isn't that fully implemented. If the script attempts to get a prompt and the method is not implemented then the pypsrp client will grind to a halt.

**Comment by malthe** [comment]
Might have something to do with `SET_PSDEBUGMODE` Message as explained here: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-psrdp/179dfdcc-4422-46cb-b662-212ec00ef572.

Setting up the `PSHost` does produce the output now from `Write-Host`. But it doesn't do anything for `Set-PSDebug`. As far as I can understand, it should be possible to get the same behavior as using `Enter-PSSession`.

I tried using the suggested `__Set-PSDebugMode` cmdlet but I get `ObjectNotFound`.

**Comment by jborean93** [comment]
> Might have something to do with SET_PSDEBUGMODE Message as explained here: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-psrdp/179dfdcc-4422-46cb-b662-212ec00ef572.

Yea that's the one, it doesn't state it should start with `__` but it's the only explanation I can think off. There are other special commands that are referenced there to set breakpoints and continue after a breakpoint is hit but obviously the client doesn't need to send them until it is needed.

> Setting up the PSHost does produce the output now from Write-Host. But it doesn't do anything for Set-PSDebug.

Yea PSHost will solve your host output but I'm unsure what your problem is with `Set-PSDebug`. Can you share what you are trying to do and what you expect to happen with this?

**Comment by malthe** [comment]
We're using JEA to restrict the shell to a predefined set of commands such as `Invoke-Job1` which would perhaps call a program `C:\Programs\Job1.exe --output-dir D:\Job1`.

Now, in JEA this looks like:
```powershell
function Invoke-Job1 {
    & C:\Programs\Job1.exe --output-dir D:\Job1
}
```
I'd like to get a debug line out to the remote client with this command-line – like [command echoing](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/echo).

The tracing capabilities are interesting as well, but the basic need is just "echo on".

**Comment by jborean93** [comment]
What would be the steps you use with `Enter-PSSession` to get the output you wish for? Just trying to get a feel for the workflow so I can replicate it myself and analyze the traffic to see what needs to be done in pypsrp.

**Comment by malthe** [comment]
I open a PowerShell on a Windows PC and use the following commands:
```powershell
C:\> Enter-PSSession -ComputerName HAL9K -UseSSL -Credential KIRK -ConfigurationName JEA1
[HAL9K]: PS> Set-PSDebug -Trace 1
[HAL9K]: PS> Invoke-Job1
```
As debugging lines, I then get the following output:
```
DEBUG: 10+ function Invoke-Job 1  >>>> {
DEBUG: 13+   >>>> & C:\Programs\Job1.exe --output-dir D:\Job1
...
```
Which is not amazing, but it can be parsed and formatted if necessary. I have tried to get the same information using `Trace-Command` or `Start-Trace` – but the trace events emitted this way need additional processing if I were to "emulate" command echoing.

**Comment by jborean93** [comment]
So looking at your output I see that a traced command does 2 things

* Outputs the `DEBUG_RECORD`
* Calls the `WriteDebugLine` on the host

You can simply just go through the debug records and process this accordingly. There's no need to do the initial `__Set-PSDebugMode` as that's used for actual breakpoints and control debugging.

```python
from pypsrp.powershell import RunspacePool, PowerShell
from pypsrp.wsman import WSMan

with WSMan(...) as wsman, RunspacePool(wsman) as rp:
    ps = PowerShell(rp)
    ps.add_script('. C:\\temp\\job.ps1; Set-PSDebug -Trace 1; Invoke-Job1')

    print(ps.invoke())

    print("DEBUG")
    print("\n".join(str(d) for d in ps.streams.debug))
```

Outputs

```
["'hi'"]
DEBUG
   1+ . C:\temp\job.ps1; Set-PSDebug -Trace 1;  >>>> Invoke-Job1
   1+ function Invoke-Job1  >>>> {

   2+      >>>> & cmd /c "echo 'hi'"

   3+  >>>> }
```

If you add the host you will also see an attempted call to `WriteDebugLine` and this is why `Enter-PSSession` displays the trace information as it comes rather than hide it in the debug stream. The `Trace-Command` cmdlet is different in that I cannot get it to output anything over a PSSession unless I add `-PSHost` to the call.  This behaviour is the same with running it normally so I don't think it's anything special with a remote PSSession. The will have the trace data be sent through in the exact same way as `Set-PSDebug -Trace 1`, e.g.

```python
from pypsrp.powershell import RunspacePool, PowerShell
from pypsrp.wsman import WSMan

with WSMan(...) as wsman, RunspacePool(wsman) as rp:
    ps = PowerShell(rp)
    ps.add_script('. C:\\temp\\job.ps1; Trace-Command ParameterBinding -Expression { Invoke-Job1 } -PSHost')

    print(ps.invoke())

    print("DEBUG")
    print("\n".join(str(d) for d in ps.streams.debug))
```

TLDR: you don't have to do anything special and no host is needed. The records are sent as a `DEBUG_RECORD` with the message being what you see displayed. The `Trace-Command` requires the `-PSHost` parameter but that's the same as a normal session.

**Comment by malthe** [comment]
Must have to do with using the `RestrictedRemoteServer` server type then – or the fact that I am using `add_cmdlet` rather than `add_script` (which I do because my session is configured with `VisibleCmdlets` to restrict the interface to a particular subset of cmdlets).

**Comment by jborean93** [comment]
That's interesting that you couldn't have it span across statements as essentially the `add_script` example in mine was doing the same thing.. It sounds like you've got a decent workaround and the problem was on the server side not respective the setting in the previous statement.
